src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs

```diff
@@ -189,18 +210,74 @@ private static void OnSocketEvent(Interop.Sys.NetworkChangeKind kind)
             }
         }
 
+        private static void OnAddressChanged()
+        {
+            Dictionary<NetworkAddressChangedEventHandler, ExecutionContext> addressChangedSubscribers = null;
+
+            lock (s_gate)
+            {
+                if (s_addressChangedSubscribers.Count > 0)
+                {
+                    addressChangedSubscribers = new Dictionary<NetworkAddressChangedEventHandler, ExecutionContext>(s_addressChangedSubscribers);
+                }
+            }
+
+            if (addressChangedSubscribers != null)
+            {
+                foreach (KeyValuePair<NetworkAddressChangedEventHandler, ExecutionContext>
+                    subscriber in addressChangedSubscribers)
+                {
+                    NetworkAddressChangedEventHandler handler = subscriber.Key;
+                    ExecutionContext ec = subscriber.Value;
+
+                    if (ec == null) // Flow supressed
+                    {
+                        handler(null, EventArgs.Empty);
+                    }
+                    else
+                    {
+                        ExecutionContext.Run(ec, s_runAddressChangedHandler, handler);
+                    }
+                }
+            }
+        }
+
         private static void OnAvailabilityTimerFired(object state)
         {
+            Dictionary<NetworkAvailabilityChangedEventHandler, ExecutionContext> availabilityChangedSubscribers = null;
+
             bool changed;
             lock (s_gate)
             {
                 changed = s_availabilityHasChanged;
                 s_availabilityHasChanged = false;
+                if (s_availabilityChangedSubscribers.Count > 0)
```

Comment:
I think this should be:
```C#
if (changed && s_availabilityChangedSubscribers.Count > 0)
```
Otherwise we're potentially allocating/copying the dictionary and then not using the copy at all.